src/pyodide/internal/_workers.py

```diff
@@ -936,44 +937,49 @@ class _WorkflowStepWrapper:
     def __init__(self, js_step):
         self._js_step = js_step
 
-    def do(self, name, config=None):
-        def decorator(func):
-            @functools.wraps(func)
-            async def wrapper(*args, **kwargs):
-                async def _callback():
-                    # Execute the user function (sync or async) and convert its result.
-                    result = func(*args, **kwargs)
-
-                    # If the result is awaitable (coroutine / Future), await it.
-                    if inspect.isawaitable(result):
-                        result = await result
-
-                    return result
-                python_callback = _callback  # `to_js` will handle the conversion.
-
-                if config is None:
-                    result = await self._js_step.do(name, python_callback)
-                else:
-                    result = await self._js_step.do(name, to_js(config), python_callback)
-
-                return result
-            return wrapper
-
-        return decorator
+    def do(self, name, callback=None, config=None):
+        if callback is None:
+            # Return a decorator that will execute the function when called
+            def decorator(func):
+                async def wrapper(*args, **kwargs):
+                    async def _callback():
+                        result = func(*args, **kwargs)
+                        if inspect.isawaitable(result):
+                            result = await result
+                        return result
+
+                    if config is None:
+                        result = self._js_step.do(name, _callback)
+                    else:
+                        result = self._js_step.do(name, to_js(config), _callback)
+
+                    return python_from_rpc(result)
+                return wrapper
+            return decorator
+
+        # Original callback-based implementation
+        async def _callback():
+            result = callback()
+            if inspect.isawaitable(result):
+                result = await result
+            return result
+
+        if config is None:
+            result = self._js_step.do(name, _callback)
+        else:
+            result = self._js_step.do(name, to_js(config), _callback)
 
-    # JS <-> PY already converts casing
-    # No further work needed for sleep/sleep_until since APIs are identical
+        return python_from_rpc(result)
```

Comment:
Let's factor this out into a separate private `_do_call()` top level function. Better not to do a method in case people define their own method with the same name on a subclass.